src/compiler/core.ts

```diff
@@ -788,6 +791,249 @@ namespace ts {
+      */
+    export function createNodeStack(): NodeStack {
+        let stackSize: number = 0;
+        let stack: Node[] = [];
+        let rootNode: Node;
+        let parentNode: Node;
+        let currentNode: Node;
+        
+        /** Gets the node at the bottom of the stack. */
+        function getRoot() {
+            return rootNode;
+        }
+        
+        /** Gets the node two steps back from the top of the stack. */
+        function getGrandparent() {
+            return peekNode(2);
+        }
+        
+        /** Gets the node one step back from the top of the stack. */
+        function getParent() {
+            return parentNode;
+        }
+        
+        /** Gets the node at the top of the stack. */
+        function getNode() {
+            return currentNode;
+        }
+        
+        /** Gets the SyntaxKind for the node at the top of the stack. */
+        function getKind() {
+            return currentNode ? currentNode.kind : undefined;
+        }
+        
+        /** Pushes a node onto the stack. */
+        function pushNode(node: Node): void {
+            stackSize++;
+            if (stackSize > 2) {
+                stack.push(parentNode);
+            }
+            else if (stackSize === 1) {
+                rootNode = node;
+            }
+            parentNode = currentNode;
+            currentNode = node;
+        }
+        
+        /** Pops the top node from the stack. */
+        function popNode(): void {
+            currentNode = parentNode;
+            parentNode = stackSize > 2 ? stack.pop() : undefined;
+            stackSize--;
+            if (stackSize === 0) {
+                rootNode = undefined;
+            }
+        }
+        
+        /** Replaces the node at the top of the stack. */
+        function setNode(node: Node): void {
+            currentNode = node;
+        }
+        
+        /** Peeks at a node a specified number of steps back from the top of the stack. */
+        function peekNode(stackOffset: number): Node {
+            switch (stackOffset) {
+                case 0: return currentNode;
+                case 1: return parentNode;
+                default: return stackSize > 2 ? stack[stackSize - 1 - stackOffset] : undefined;
+            }
+        }
+
+        /** Traverses the stack from top to bottom until it finds a node that matches the supplied predicate. */
+        function findAncestorNode<T extends Node>(match: (node: Node) => node is T): T;
+        /** Traverses the stack from top to bottom until it finds a node that matches the supplied predicate. */
+        function findAncestorNode(match: (node: Node) => boolean): Node;
+        function findAncestorNode(match: (node: Node) => boolean) {
+            if (parentNode && match(parentNode)) {
+                return parentNode;
+            }
+            for (let i = stack.length; i >= 0; i--) {

```

Comment:
No, `stack` only contains grandparent and higher nodes, to avoid array operations for common use cases. We already checked the parent node above, so to check other ancestors we check every element of `stack` starting from the top. This should, however, be `stack.length - 1`.